File: src/log/statement/StatementBuilder.java
package log.statement;

import java.util.Collection;

/**
 * Permet de construire une proposition à partir de valeurs quelconques,
 * en gérant de facon simplifiée, les éventuelles imbrications possibles.
 */
public class StatementBuilder {
    // ATTRIBUTS
    /** La proposition principale, en cours de construction **/
    private final ComposedStatement main;


    // CONSTRUCTEUR
    public StatementBuilder() {
        main = new ComposedStatement();
    }


    // REQUÊTES
    /** Renvoie la proposition associée à ce constructeur **/
    public Statement toStatement() {
        return main;
    }

    /** Renvoie le texte correspondant à la proposition construite, jusqu'à présent */
    public String toString() {
        return main.toString();
    }

    // COMMANDES
    /**
     * Rajoute un nouvel élément simple à la proposition (sous sa représentation texte)
     *
     * @param e Le nouvel élément à ajouter à la proposition
     */
    public <E> void append(E e) {
        main.add(new SimpleStatement(e));
    }

    /**
     * Rajoute une liste d'éléments, à faire correspondre sous forme d'une unique proposition
     * (énumérations de valeurs)
     *
     * @param collection Les éléments à énumérer dans une sous-proposition
     * @param separator La chaîne séparant les différents éléments de la collection.
     */
    public <E> void append(Collection<E> collection, String separator) {
        assert collection != null;
        assert separator != null;

        ComposedStatement cs = new ComposedStatement(separator); {
            for (E e : collection) {
                cs.add(new SimpleStatement(e));
            }
        }

        main.add(cs);
    }
}

File: src/main/statement/StatementBuilder.scala
package main.statement

/**
  * Permet de construire une proposition à partir de valeurs quelconques,
  * en gérant de facon simplifiée, les éventuelles imbrications possibles.
  */
class StatementBuilder extends IStatementBuilder {
  /** La proposition principale, en cours de construction **/
  private val main: ComposedStatement = new ComposedStatement("")

  // REQUETES
  /** Renvoie la proposition associée à ce constructeur **/
  override def toStatement: Statement = main

  /** Renvoie le texte correspondant à la proposition construite, jusqu'à présent */
  override def toString: String = main.toString

  // COMMANDES
  /**
    * Rajoute un nouvel élément simple à la proposition (sous sa représentation texte)
    *
    * @param e Le nouvel élément à ajouter à la proposition
    */
  override def append[E](e: E) {
    main.add(new SimpleStatement(e))
  }

  /**
    * Rajoute une liste d'éléments, à faire correspondre sous forme d'une unique proposition
    * (énumérations de valeurs)
    *
    * @param collection Les éléments à énumérer dans une sous-proposition
    * @param separator  La chaîne séparant les différents éléments de la collection.
    */
  override def append[E](collection: Seq[E], separator: String) {
    assert(collection != null)
    assert(separator != null)

    val test = Seq.empty
    val cs: ComposedStatement = new ComposedStatement(separator) {
      collection.foreach((e: E) => cs.add(new SimpleStatement(e)))
    }

    main.add(cs)
  }
}